### ISSUE TITLE
Output Discrepancy in BatchDecodeWithPagedKVCacheWrapper with use_tensor_cores=True vs. False

### ISSUE THREAD

**Issue by KanishkGoel00** [flashinfer-ai/flashinfer#703]
#### Summary:
I observed a discrepancy in the output of `BatchDecodeWithPagedKVCacheWrapper` when the `use_tensor_cores` flag is set to `True` versus `False`. Despite identical inputs and configurations, the outputs differ slightly. Below are the test results, including latency and relative differences. Is this level of discrepancy expected when toggling use_tensor_cores? Is this related to the type discrepancy mentioned [here](https://github.com/flashinfer-ai/flashinfer/issues/520#issuecomment-2430183276)?

#### Test Configuration:
- **Batch size (`bs`)**: 57  
- **Sequence length (`cl`)**: 8192  
- **Number of attention heads (`num_heads`)**: 32  
- **Number of key-value heads (`num_kv_heads`)**: 8  
- **Head dimension (`head_dim`)**: 128  
- **Block size (`block_size`)**: 16  
- **Random seed**: 42 (to ensure reproducibility)  
- Comparison between `use_tensor_cores=True` and `use_tensor_cores=False`

#### Results:
1. **Latency:**
   - `use_tensor_cores=True`: `1.276 ms`
   - `use_tensor_cores=False`: `1.595 ms`
   - Latency ratio (`use_tensor_cores=False` / `use_tensor_cores=True`): `1.25`

2. **Input Matching:**
   - Query (`q`) inputs match: **Yes**  
   - Key-Value (`kv_data`) inputs match: **Yes**  

3. **Output Matching:**  
   - Outputs match: **No**  

   Differences in outputs:  
   - `max_abs_diff`: 0.000092  
   - `mean_abs_diff`: 0.000008  
   - `rmsd`: 0.000000  
   - `max_rel_diff`: 1.579102  
   - `mean_rel_diff`: 0.002205  

#### Environment:
- **PyTorch version**: 2.5.1+cu118  
- **CUDA available**: True  
- **CUDA version**: 12.2 
- **GPU**: NVIDIA A100 80GB PCIe  
- **FlashInfer version**: 0.1.6+cu121torch2.4  
- **Operating System**: Linux 6.2.0
- **Python version**: 3.10.16  

#### Code:
Here is the code used to generate the above results:

```python
import torch
import math
import flashinfer as fi  # Assuming fi contains BatchDecodeWithPagedKVCacheWrapper

def compare_tensors(tensor1, tensor2, tolerance=1e-5):
    match = torch.allclose(tensor1, tensor2, atol=tolerance, rtol=tolerance)
    differences = {}
    if not match:
        denom = torch.where(tensor1.abs() > tolerance, tensor1, torch.ones_like(tensor1))
        differences["max_abs_diff"] = torch.max(torch.abs(tensor1 - tensor2)).item()
        differences["mean_abs_diff"] = torch.mean(torch.abs(tensor1 - tensor2)).item()
        differences["rmsd"] = torch.sqrt(torch.mean((tensor1 - tensor2) ** 2)).item()
        differences['max_rel_diff'] = torch.max(torch.abs((tensor1 - tensor2) / denom)).item()
        differences['mean_rel_diff'] = torch.mean(torch.abs((tensor1 - tensor2) / denom)).item()
    return match, differences

@torch.inference_mode
def do_flashinfer_decode_paged(bs, cl, num_heads, num_kv_heads, head_dim, block_size, use_tensor_cores, seed=None):
    if seed is not None:
        torch.manual_seed(seed)
        torch.cuda.manual_seed_all(seed)
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    dtype = torch.float16
    warmup_steps = 0
    active_steps = 1
    try:
        q = torch.randn(bs, num_heads, head_dim, dtype=dtype, device=device)
        workspace_buffer = torch.empty(32 * 1024 * 1024, dtype=torch.int8, device=device)
        decode_wrapper = fi.BatchDecodeWithPagedKVCacheWrapper(workspace_buffer, "NHD", use_tensor_cores=use_tensor_cores)
        num_pages_per_req = math.ceil(cl / block_size)
        max_num_pages = num_pages_per_req * bs
        kv_page_indices = torch.arange(max_num_pages, dtype=torch.int32, device=device)
        kv_page_indptr = torch.arange(0, bs + 1, dtype=torch.int32, device=device) * num_pages_per_req
        kv_last_page_len = torch.full((bs,), (cl - 1) % block_size + 1, dtype=torch.int32, device=device)
        kv_data = torch.randn(max_num_pages, 2, block_size, num_kv_heads, head_dim, dtype=dtype, device=device)
        decode_wrapper.begin_forward(kv_page_indptr, kv_page_indices, kv_last_page_len, num_heads, num_kv_heads, head_dim, block_size)
        for _ in range(warmup_steps):
            decode_wrapper.forward(q, kv_data)
        start = torch.cuda.Event(enable_timing=True)
        end = torch.cuda.Event(enable_timing=True)
        start.record()
        for _ in range(active_steps):
            decode_wrapper.forward(q, kv_data)
        end.record()
        torch.cuda.synchronize()
        elapsed_time_ms = start.elapsed_time(end) / active_steps
        output = decode_wrapper.forward(q, kv_data)
        return elapsed_time_ms, output, q, kv_data
    except Exception as e:
        print(f"Error during FlashInfer decode (paged) profiling: {e}")
        return -1, None, None, None

# Example usage
seed = 42
bs, cl, num_heads, num_kv_heads, head_dim, block_size = 57, 8192, 32, 8, 128, 16
latency_tc, output_tc, q_tc, kv_data_tc = do_flashinfer_decode_paged(bs, cl, num_heads, num_kv_heads, head_dim, block_size, True, seed)
latency_ntc, output_ntc, q_ntc, kv_data_ntc = do_flashinfer_decode_paged(bs, cl, num_heads, num_kv_heads, head_dim, block_size, False, seed)
inputs_match_q, q_diff = compare_tensors(q_tc, q_ntc)
inputs_match_kv, kv_diff = compare_tensors(kv_data_tc, kv_data_ntc)
outputs_match, output_diff = compare_tensors(output_tc, output_ntc)

print("Latency with use_tensor_cores=True:", latency_tc)
print("Latency with use_tensor_cores=False:", latency_ntc)
print("Inputs (q) match:", inputs_match_q)
print("Inputs (kv_data) match:", inputs_match_kv)
print("Outputs match:", outputs_match)
if not outputs_match:
    print("Differences in outputs:")
    for metric, value in output_diff.items():
        print(f"  {metric}: {value:.6f}")

**Comment by yzh119** [comment]
Yes decode kernel uses fp32 for compute internally.

Another possible reason for output discrepancy is that we have to round $P$ to f16 for the second GEMM ($P\cdot V$) in the prefill attention kernel, which might cause some precision loss.

**Comment by KanishkGoel00** [comment]
Thank you for your earlier response and clarification regarding the output discrepancies due to the internal use of FP32. I appreciate the insights.

I have observed another interesting point related to the prefill and decode attention kernels, particularly in terms of resource utilization (registers and shared memory) per threadblock. Below are the launch statistics reported using ncu for both kernels when toggling `use_tensor_cores`:

**Prefill Attention Kernel (`use_tensor_cores=True`)**
- Registers Per Threadblock: 182
- Shared Memory (Dynamic) Per Threadblock: 69.63 KB
- Shared Memory (Static) Per Threadblock: 0
- Waves Per SM: 2.11

**Decode Attention Kernel (`use_tensor_cores=False`)**
- Registers Per Threadblock: 66
- Shared Memory (Dynamic) Per Threadblock: 9.22 KB
- Shared Memory (Static) Per Threadblock: 0
- Waves Per SM: 0.60

### Observations:
1. **Registers Usage:**
   - The prefill attention kernel uses significantly more registers per threadblock (182 vs. 66). 
2. **Shared Memory Usage:**
   - Prefill kernel uses much higher dynamic shared memory per threadblock (69.63 KB vs. 9.22 KB).

### Question:
Is the significantly higher register and shared memory usage for the prefill attention kernel a direct result of tensor cores being used? Alternatively, could these values be optimized further to reduce per-threadblock register and shared memory usage while maintaining FP16/BF16 precision? Specifically, I’m curious if optimizations for tensor core utilization inherently led to higher resource usage per threadblock.

Thank you for looking into this, and I look forward to your insights!

**Comment by yzh119** [comment]
Register and shared memory usage are determined by tile sizes and head dimensions `(CTA_TILE_Q, CTA_TILE_KV, HEAD_DIM)`.

# Register usage

We store the first GEMM's output ($P$) and second GEMM's output ($O$) in registers (there are other elements that have to be stored in registers but these two are dominant), both in f32, the number of registers per thread is determined by:
```
#REGISTERS_PER_THREAD(P) = CTA_TILE_Q * CTA_TILE_KV / NUM_THREADS
#REGISTERS_PER_THREAD(O) = CTA_TILE_Q * HEAD_DIM / NUM_THREADS
```

# Shared Memory Usage

We store the query tile $Q$ and kv tile $K$/ $V$ in shared memory, shared memory usage is determined by:
```
#SMEM_SIZE(Q) = CTA_TILE_Q * HEAD_DIM * sizeof(DTypeQ)
#SMEM_SIZE(KV) = 2 * CTA_TILE_K * HEAD_DIM * sizeof(DTypeKV) * #PIPELINE_DEPTH
```

For decode kernels, we use cuda cores and the `CTA_TILE_Q` is set to `num_qo_heads / num_kv_heads` (for grouped-query attention). For prefill kernels, we use Tensor Cores, and we pad `query_len * num_qo_heads / num_kv_heads` to a query tile size that can fit into tensor cores, the minimal `CTA_TILE_Q` is 16, however, for most decoding case (`query_len=1`), `num_qo_heads / num_kv_heads` cannot reach this value so you can observe prefill kernels uses much more registers than decode kernels (one of the future optimizations we can do for decoding is to support `CTA_TILE_Q` of 8 but still uses Tensor Cores with $M=16$, we only allocate half of the registers, and pass placeholder for the lower half matrix in tensor cores instructions).

**Comment by KanishkGoel00** [comment]
Understood, thanks for the prompt responses!
